### ISSUE TITLE
Effective Instant serialization/deserialization

### ISSUE THREAD

**Issue by egnevushev** [nodatime/nodatime#1644]
I've faced a problem with effective serialization and deserialization `Instant` for MessagePack. `Instant` has a precision of nanoseconds. The best safe way for serialization is getting nanoseconds amount (and create an instance of `Instant` from nanoseconds for deserialization). But the minimum accessible unit of time from Instant is ticks (`.ToUnixTimeTicks()`). 
I can't use ticks, because value can be truncated (`Instant.MaxValue` for example)

Now there are a few terrible ways to serialize/deserialize an Instant:
- with InstantPattern (as string with format pattern). It's a slowest way, and uses a lot of memory
- with code smells way: getting `Duration` with `Instant.MinValue - instant` for serialization, and getting `Instant` with 'Instant.Add(Instant.MinValue, duration)' for deserialization
- with dirty hacks for getting access to private fields: Reflection, Linq Expressions

I would like to propose change access modifier for `Duration` field to public. And make a constructor with `Duration` as argument public. It will help with serialization and make `Instant` more flexible and predictable with nanoseconds manipulating.

Do you have any thoughts about this?

**Comment by jskeet** [comment]
I would *definitely* not make any field public. That just breaks encapsulation.

I also don't want to make a constructor that accepts a `Duration` - because that makes an *assumption* of a particular epoch, which I've been trying to avoid. (We used to have that in 1.x, and I worked hard to remove it.) We should be able to change the internal epoch without breaking anything.

I think it's entirely reasonable for you to serialize an instant by determining some duration from *whatever epoch you want* (e.g. `Instant.MinValue`, `NodaConstants.UnixEpoch` or `NodaConstants.BclEpoch`) using normal calculations, and perform the calculation. I don't see how that's "terrible".

**Comment by Sasha654** [comment]
Hi!
What do you think of extending the API for `Instant` by adding the ToUnixTimeNanoseconds and FromUnixTimeNanoseconds(Int64) methods?

**Comment by jskeet** [comment]
While I'd be okay with that, I wouldn't expect it to be used by serialization frameworks - as it would only cover instants for years in the range of around 1678 to 2262. That's fairly limiting, even if it would be fine for quite a few scenarios.

**Comment by Sasha654** [comment]
Yes, this range is good in many ways, including for this case. You can also go ahead and add more like `ToBigIntegerNanoseconds` and `FromNanoseconds(BigInteger) `

**Comment by jskeet** [comment]
I certainly could do - but I'm not sure that it's *much* better than `NodaConstants.UnixEpoch + Duration.FromNanoseconds(nanos)` or `(instant - NodaConstants.UnixEpoch).ToBigIntegerNanoseconds()`.

I've generally tried to be reasonably conservative with the public API, and this is only a small additional convenience over what you can do now. In particular, if this is going to be part of a serialization framework, presumably that slightly-less-convenient code only needs to be written once.

**Comment by Sasha654** [comment]
Anyway, how about I make a pull request to at least To/From nanos for `Instant`?

**Comment by jskeet** [comment]
I would rather discuss it first. The implementation would be trivial - the cost isn't the implementation, it's making the public API larger for relatively little benefit. (But in neither case would it be Instant.ToNanoseconds or Instant.FromNanoseconds - it would Instant.ToUnixEpochNanoseconds and Instant.FromUnixEpochNanoseconds.)

**Comment by Sasha654** [comment]
Ok. Having access to nanoseconds in all cases is good, even without considering the purpose of the serialization.
Of course it would be Instant.ToUnixTimeNanoseconds and Instant.FromUnixTimeNanoseconds.

**Comment by jskeet** [comment]
> Having access to nanoseconds in all cases is good, even without considering the purpose of the serialization.

Users already "have access to nanoseconds in all cases".

These methods would do nothing that can't be done already. Here's their implementation:

```csharp
public static Instant FromUnixTimeNanoseconds(long nanoseconds) => NodaConstants.UnixEpoch + Duration.FromNanoseconds(nanoseconds);
public static Instant FromUnixTimeNanoseconds(BigInteger nanoseconds) => NodaConstants.UnixEpoch + Duration.FromNanoseconds(nanoseconds);
public long ToUnixTimeInt64Nanoseconds() => duration.ToInt64Nanoseconds();
public BigInteger ToUnixTimeBigIntegerNanoseconds() => duration.ToBigIntegerNanoseconds();
```

The last two methods do use duration, but are absolutely equivalent to:

```csharp
public long ToUnixTimeInt64Nanoseconds() => (this - NodaConstants.UnixEpoch).ToInt64Nanoseconds();
public BigInteger ToUnixTimeBigIntegerNanoseconds() => (this - NodaConstants.UnixEpoch).ToBigIntegerNanoseconds();
```

The "native" implementation would be very slightly more efficient, but I'd be surprised if it made a huge difference in any real-world scenario. In other words, I don't expect these new methods to *unblock* anything.

**Comment by Sasha654** [comment]
However, the proposed API extension for Instant will not break consistency, and getting nanoseconds directly from Instant without additional manipulation will be more convenient.

**Comment by jskeet** [comment]
I agree it won't break consistency. But that doesn't mean it doesn't have a cost in terms of cognitive load. I note that we've had nanosecond precision for over 4 years now, and I *think* this is the first time it's been requested... so I expect it to be used by very few people. Whereas everyone hitting "." in Visual Studio will see it *offered* to them alongside the other methods.

Anyone who wants the clarity of calling code can add extension methods for the moment for `ToUnixTime{...}Nanoseconds`, while we continue to consider this. There's admittedly no equivalent to extension methods for the `FromUnixTimeNanoseconds` static methods, but it's at least not hard to call `InstantHelper` etc.

Basically: I don't see this as *urgent* and I'd like to spend some time considering it and seeing if there's more demand for it, before adding it.

**Comment by ptupitsyn** [comment]
I have faced a similar thing with `Instant` serialization. I'm developing a .NET driver for a Java-based DB (Apache Ignite). Java side uses `java.time.Instant` and writes it like this:

```
long seconds = value.getEpochSecond();
int nanos = value.getNano();
putLong(seconds);
if (nanos != 0) {
    putInt(nanos);
}
```

Deserializing this in .NET is easy: `Instant.FromUnixTimeSeconds(seconds).PlusNanoseconds(nanos)`

Serialization does not sit right with me though: 
```
long seconds = value.ToUnixTimeSeconds();
int nanos = (value - NodaConstants.UnixEpoch).SubsecondNanoseconds;
```

It would be great to have `Instant.SubsecondNanoseconds` property to make this "symmetric".

**Comment by jskeet** [comment]
`Instant.SubsecondNanoseconds` is epoch-specific though, and I've tried very hard to avoid exposing the epoch *implicitly*. (It's fine to have methods that explicitly mention the Unix epoch.) Subtracting the unix epoch to get a duration is the right way of doing this, IMO. The way I look at it, what you're serializing is actually a duration - and the expectation is that you can add that duration to the Unix epoch in order to get an instant.

Ideally, I'd prefer to see the code take that approach - use `value - NodaConstants.UnixEpoch`, and then get the seconds and nanoseconds separately. Currently there isn't a clear way of doing that though, I'll readily admit.

I believe your current code has a bug for values before the Unix epoch that aren't on "second boundaries": `ToUnixTimeSeconds()` will round towards the start of time, but `SubsecondNanoseconds` will give a negative value.

So for 1.25 seconds before the Unix epoch, you'd end up with seconds=-2, nanoseconds=-250,000,000 - where presumably you'd *want* seconds=-2, nanoseconds=750,000,000.

This is the equivalent code in `NodaTime.Serialization.Protobuf`:

```csharp
// Truncated towards the start of time, which is what we want...
var seconds = instant.ToUnixTimeSeconds();
var remainder = instant - Instant.FromUnixTimeSeconds(seconds);
// NanosecondOfDay is probably the most efficient way of turning a small, subsecond, non-negative duration
// into a number of nanoseconds...
return new Timestamp { Seconds = seconds, Nanos = (int) remainder.NanosecondOfDay };
```

I do wonder whether (seconds, nanoseconds) where nanoseconds is always positive is a sufficiently common representation that I would bake that into Noda Time directly - I wouldn't be averse to `Instant.ToUnixTimeSecondsAndNanoseconds` for example. I suspect I could make that more efficient.

**Comment by ptupitsyn** [comment]
@jskeet thanks for the pointer to `NodaTime.Serialization.Protobuf`, very helpful!
